src/main/kotlin/com/exactpro/th2/codec/grpc/GrpcCodecService.kt

```diff
@@ -34,52 +40,56 @@ class GrpcCodecService(
-        .connectionId.sessionAlias
+    private val MessageGroupBatch.sessionAlias get() = getGroups(0)
+        .getMessages(0).sessionAlias
+        .also { if (!checkIfSameSessionAlias(it)) eventProcessor.onErrorEvent("Batch contains messages with different session aliases.") }
+
+    private fun MessageGroupBatch.checkIfSameSessionAlias(alias: String) =
+        groupsList.all { groups -> groups.messagesList.all { it.sessionAlias == alias } }
+
+    private val AnyMessage.sessionAlias get() = if (hasMessage()) message.sessionAlias else rawMessage.sessionAlias
 
     override fun decode(message: MessageGroupBatch, responseObserver: StreamObserver<MessageGroupBatch>) {
         try {
             val parsed = generalDecodeFunc(message)
-            if (nextCodec != null && parsed.anyMessage(AnyMessage::hasRawMessage) ) {
-                nextCodec.decode(parsed, mapOf(SESSION_ALIAS_LABEL to getSessionAlias(parsed)), responseObserver)
+            if (parsed.anyMessage(AnyMessage::hasRawMessage)) {
+                if (nextCodec == null) {
+                    eventProcessor.onErrorEvent("codec pipeline output contains raw messages")
+                } else {
+                    nextCodec.decode(parsed, mapOf("session_alias" to parsed.sessionAlias), responseObserver)
+                }
             } else {
                 responseObserver.onNext(parsed)
                 responseObserver.onCompleted()
             }
         } catch (t: Throwable) {
+            eventProcessor.onErrorEvent("'decode' rpc call exception", cause = t)
             LOGGER.error(t) { "'decode' rpc call exception" }
-            responseObserver.onError(t)
+            responseObserver.onError(Status.INTERNAL.withDescription(t.message).withCause(t).asException())
         }
     }
 
     override fun encode(message: MessageGroupBatch, responseObserver: StreamObserver<MessageGroupBatch>) {
-        val nextCodecObserver = object : StreamObserver<MessageGroupBatch> {
-            override fun onNext(value: MessageGroupBatch) {
-                val raw = generalEncodeFunc(value)
-                responseObserver.onNext(raw)
-            }
-            override fun onError(t: Throwable) = responseObserver.onError(t)
-            override fun onCompleted() = responseObserver.onCompleted()
+        val nextCodecObserver = object : StreamObserver<MessageGroupBatch> by responseObserver {
+            override fun onNext(value: MessageGroupBatch) = responseObserver.onNext(generalEncodeFunc(value))
         }
 
         try {
-            if (nextCodec != null && message.anyMessage(AnyMessage::hasMessage) ) {
-                nextCodec.encode(message, mapOf(SESSION_ALIAS_LABEL to getSessionAlias(message)), nextCodecObserver)
+            if (nextCodec != null && message.anyMessage(AnyMessage::hasMessage)) {
+                nextCodec.encode(message, mapOf("session_alias" to message.sessionAlias), nextCodecObserver)
```

Comment:
We should produce an error if there's no next codec but there are still messages to process in a batch - as we already do in decode

Reply:
During encoding pipeline works in opposite direction - from last to first. Last codec receives unprocessed batch and this is correct. We should produce an error in first codec ( in responseObserver.onNext() ), but we don't know if current codec is first in pipeline or not.